Update and delete (#8)
Ada-Developers-Academy/core-unit-2: update-and-delete/update-and-delete.md

| `db.session.commit()`                               | Every time a SQLAlchemy model has been updated, and we want to commit the change to the database, we'll execute `db.session.commit()`.     |
| `return Response( ... )`                            | This is one of many ways we can return our appropriate HTTP response                                                                       |

### !callout-info

## Naming `form_data`

We've named the variable that holds our HTTP request body `form_data`. This name helps evoke the story of this endpoint; it's likely that a client makes this request because a user filled out an update form. Other names are appropriate, though, such as `request_body`.

### !end-callout

### Manually Testing with Postman

We can use Postman to check our update functionality.

It may help to first make a request to `GET /books/1` before updating it.

![Screenshot of Postman featuring a request of GET /books/1 and a response of 200 with book data](../assets/update-and-delete/update-and-delete_update-precheck-get.png)

> Then, make a request to `PUT /books/1` with our proper JSON request body.

Author: This is such a small suggestion, I recognize I might be fixating on it, but I don't feel like this is necessary, and I would love reassurance that this will help the student experience!

![Screenshot of Postman featuring a request of PUT /books/1 and a response of 200 with a success message](../assets/update-and-delete/update-and-delete_update-request.png)

We can send another request to `GET /books/1` to confirm the updates worked.

![Screenshot of Postman featuring a request of GET /books/1 and a response updated book data](../assets/update-and-delete/update-and-delete_update-confirm-get.png)

### Manually Testing with `psql`

> We can also check if our book was updated in the database using `psql`!

Reviewer: flesh this out a bit as in the last lesson (at least for this first one)
```suggestion
We can also check if our book was updated in the database using `psql`, just as we did when creating new records!

Let's try it out on our own! We can refer back to the Intro to SQL lessons if we need a reminder about how to write a query like this.
```

<!-- prettier-ignore-start -->
### !challenge
* type: checkbox
* id: pe3GBP
* title: Update and Delete, Updating a Book Endpoint
##### !question

Think about the "Updating a Book Endpoint."

Check off all the topics that we've briefly touched on so far.

##### !end-question
##### !options

> * Predicted the HTTP response, request, and logic for this endpoint

Reviewer: match lesson changes
```suggestion
* Planned the HTTP response, request, and logic for this endpoint
```

* Refactored our `"/books/<book_id>"` endpoint to accept `PUT` requests
* Used `Book.query.get(book_id)` to get an instance of `Book` matching `book_id`
* Refactored our code to conditionally check the request's method
* Read the HTTP request body using `request.get_json()`

> * Saved the instance of `Book` in the database

Reviewer: reflect the actual action
```suggestion
* Updated the instance of `Book` in the database
```

* Returned a response
* Tested this request in Postman

##### !end-options
##### !answer

> * Predicted the HTTP response, request, and logic for this endpoint

Reviewer: match the question revision
```suggestion
* Planned the HTTP response, request, and logic for this endpoint
```

* Refactored our `"/books/<book_id>"` endpoint to accept `PUT` requests
* Used `Book.query.get(book_id)` to get an instance of `Book` matching `book_id`
* Refactored our code to conditionally check the request's method
* Read the HTTP request body using `request.get_json()`

> * Saved the instance of `Book` in the database

Reviewer: match the question revision
```suggestion
* Updated the instance of `Book` in the database
```

* Returned a response
* Tested this request in Postman

##### !end-answer
### !end-challenge
<!-- prettier-ignore-end -->

### !callout-warning

## What About Error Handling?

> There are many cases that weren't covered in this lesson, even though they are relevant to creating and reading `Book`s. For example, what happens if we make a `GET` request to `/books/this-book-doesnt-exist`? We are intentionally not covering these cases at the moment, to limit this lesson. However, hypothesize and research how to handle erroneous HTTP requests. Follow your curiosity!

Reviewer: match the verb
```suggestion
There are many cases that weren't covered in this lesson, even though they are relevant to creating and reading `Book`s. For example, what happens if we make a `PUT` request to `/books/this-book-doesnt-exist`? We are intentionally not covering these cases at the moment, to limit this lesson. However, hypothesize and research how to handle erroneous HTTP requests. Follow your curiosity!
```

### !end-callout

## Deleting a Book Endpoint: Preparation

Let's consider how to accomplish this feature:

> As a client, I want to send a request to delete one existing book and get a success response, so that I know the API deleted the book data.

> ### Predict HTTP Requests, Responses, and Logic

Reviewer: ```suggestion
### Planning HTTP Requests, Responses, and Logic
```

> Let's consider the endpoint to create a book:

Reviewer: prompt for verb and endpoint consideration
```suggestion
We should think about the typical HTTP verb and endpoint used for requests that delete a particular model record.

For this feature, we should make a `DELETE` request to the `/books` path, and we'll need to include the `id` of the record to delete as part of the endpoint.
```

| HTTP Method | Endpoint   |
| ----------- | ---------- |
| `DELETE`    | `/books/1` |

> No request body.

Reviewer: details about why we don't need the request body
```suggestion
`DELETE` requests do not generally include a request body, so no additional planning around the request body is needed.
```

This assumes that there is at least a `book` table with the following row:

> | `id` | `title`                        | `description`                                     |

Reviewer: formatting
```suggestion
| <div style="min-width:40px;">`id`</div> | `title`                | `description`                               |
```

| ---- | ------------------------------ | ------------------------------------------------- |
| `1`  | `Updated Fictional Book Title` | `This fantasy novel has robots and pirates, too.` |

> The response we want to send back is:

Reviewer: discussion of the response
```suggestion
When the record is successfully updated, we should return the appropriate status code, which is `200 OK`. Some APIs return the the record that was just deleted as the response body, but for simplicity, we will return a brief status message.
```

| Response Status | Response Body                  |
| --------------- | ------------------------------ |
| `200 OK`        | `Book #1 successfully deleted` |

> During this method, we will need to:

Reviewer: typical transition text
```suggestion
Now that we have an idea of what our endpoint should look like, we can turn our attention to how to implement it.

Our endpoint will need to:
```

> 1. Read the new, updated book data from the HTTP request
> 1. Find the instance of `Book` that has the existing book data
> 1. Delete this instance of `Book`
> 1. Save this change in the db
> 1. Send back a response

Reviewer: call out using the route param and looking it up in the db
characterize the `delete` call as informing the database of our intent to delete, prior to the commit call

```suggestion
1. Read the `book_id` in the request path
1. Retrieve the `Book` instance with the matching `book_id` from the database
1. Tell the database to delete this instance of `Book`
1. Save this change in the database
1. Send back a response
```

Author: Just want to explicitly say that I like this a lot and I'm learning form it!

## Deleting a Book Endpoint: Code

> Let's expand our `book` endpoint.
> 
> Let's look at this example code for our delete feature.

Reviewer: similar updates as above
```suggestion
This endpoint uses the same path as our existing parameterized route, `"/<book_id>"`, so we can refactor and expand on the same function.

Let's modify our endpoint code to support the delete feature.
```

```python
@books_bp.route("/<book_id>", methods=["GET", "PUT", "DELETE"])
def book(book_id):

>     book = Book.query.get(book_id)

Reviewer: as mentioned above, we should consider not having a variable shadow the function it's defined in

    if request.method == "GET":
        # ... existing code for getting a single book
    elif request.method == "PUT":
        # ... existing code for updating a single book
    elif request.method == "DELETE":
        db.session.delete(book)
        db.session.commit()

>         return Response(f"Book #{book.id} successfully deleted", status=200)

Reviewer: ```suggestion
        return make_response(f"Book #{book.id} successfully deleted")
```

```

| <div style="min-width:250px;"> Piece of Code </div> | Notes                                                                                                                                                 |
| --------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------- |
| `methods=["GET", "PUT", "DELETE"]`                  | We must update this endpoint so it accepts `DELETE` requests                                                                                          |

> | `book = Book.query.get(book_id)`                    | This feature will need to find our matching `Book` instance, so we still need to use `book`                                                           |

Reviewer: remind what book is and what this line does to fill it
```suggestion
| `book = Book.query.get(book_id)`                    | This feature will need to find our matching `Book` instance, so we still need the lookup code using `book_id` and store the found book in `book`                                                           |
```

| `elif request.method == "DELETE":`                  | We can add a new conditional branch for `DELETE` requests                                                                                             |

> | `db.session.delete(book)`                           | We can use SQLAlchemy's functions to delete our `book` with `db.session.delete(book)`                                                                 |

Reviewer: explain how this is notifying sqlalchemy of the pending deletion
```suggestion
| `db.session.delete(book)`                           | We can use SQLAlchemy's functions to tell the database to prepare to delete our `book` with `db.session.delete(book)`                                                                 |
```

| `db.session.commit()`                               | We use this function to actually apply our database changes                                                                                           |

> | `return Response(...)`                              | This is one of many ways to return an HTTP response                                                                                                   |

Reviewer: ```suggestion
| `return make_response(...)`                              | This is one of many ways to return an HTTP response. Since we didn't supply a status code, Flask will default to `200 OK`.                                                                                                   |
```

Author: This specific suggestion is helping me remember to continue to be repetitive in this curriculum (which I put a lot of effort into in Unit 1, but probably slipped on during Unit 2. Good to refresh myself for Unit 3!)

> | `f"Book #{book.id} successfully deleted"`           | We can still access `book.id` because the variable itself is still in scope in our app, even if the `book` _was_ successfully deleted in the database |

Reviewer: explicitly state that retrieved book instance isn't a live copy attached to changes in the database
```suggestion
| `f"Book #{book.id} successfully deleted"`           | We can still access `book.id` because the variable itself is still in scope in our app, even if the `book` _was_ successfully deleted in the database. The object stored in memory and referenced by `book` does not automatically update in response to changes in the database. |
```

Author: Let's try it this cohort without the callout, and I'll look out for how that goes

### Manually Testing in Postman

After writing this code, we can use Postman to test our delete functionality.

First, let's make a `GET` request to `/books/1` to confirm that a book with ID 1 exists.

![Screenshot of Postman featuring a request of GET to /books/1 and a response of 200 with book data](../assets/update-and-delete/update-and-delete_delete-get-book.png)

Then, let's make our `DELETE` request to `/books/1` and check the response.

![Screenshot of Postman featuring a request of DELETE to /books/1 and a response of 200 with a success message](../assets/update-and-delete/update-and-delete_delete-request.png)

Afterwards, let's even make a `GET` request to `/books`. We see that there are no more books that exist! Our book was successfully deleted.

![Screenshot of Postman featuring a request of GET to /books and a response of 200 with an empty array](../assets/update-and-delete/update-and-delete_delete-get-books-empty.png)

> We could also use `psql` to check this, too.

Reviewer: similar (but smaller) expansion as above
```suggestion
### Manually Testing with `psql`

We can also check if our book was deleted from the database using `psql`.

Let's try it out on our own!
```

<!-- prettier-ignore-start -->
### !challenge
* type: checkbox
* id: wYm5yT
* title: Update and Delete, Deleting a Book Endpoint
##### !question

Think about the "Deleting a Book Endpoint."

Check off all the topics that we've briefly touched on so far.

##### !end-question
##### !options

> * Predicted the HTTP response, request, and logic for this endpoint

Reviewer: ```suggestion
* Planned the HTTP response, request, and logic for this endpoint
```

* Refactored our `"/books/<book_id>"` endpoint to accept `DELETE` requests
* Used `Book.query.get(book_id)` to get an instance of `Book` matching `book_id`
* Refactored our code to conditionally check the request's method
* Deleted the `book` from the database
* Committed these changes in the database
* Returned a response
* Tested this request in Postman

##### !end-options
##### !answer

> * Predicted the HTTP response, request, and logic for this endpoint

Reviewer: ```suggestion
* Planned the HTTP response, request, and logic for this endpoint
```
